---
title: "Lab 11: Structural Bioinformatics 2"
author: "Nicholas Chiu"
format: pdf
---

## Lab 10 Completion

Lab 10, including PCA, was submitted in full in the Lab 10 gradescope assignment

# Alpha Fold

## 6. Generating your own structure predictions

Colabfold plots:

![colabfold plot 1](alphafold1.png)

![colabfold plot 2](alphafold2.png)

## 7. Interpreting Results

![molstar plot 1](molstar1.png)

![molstar plot 2](molstar2.png)

## 8.1 Custom analysis of resulting models

```{r}
results_dir <- "hivprdimer_23119/" 


# File names for all PDB models
pdb_files <- list.files(path=results_dir, pattern="*.pdb", full.names = TRUE)

# Print our PDB file names
basename(pdb_files)

```

```{r}
library(bio3d)

# Read all data from Models 
#  and superpose/fit coords
pdbs <- pdbaln(pdb_files, fit=TRUE, exefile="msa")

pdbs
```

```{r}
rd <- rmsd(pdbs, fit=T)

range(rd)

#install.packages("pheatmap")
library(pheatmap)

colnames(rd) <- paste0("m",1:5)
rownames(rd) <- paste0("m",1:5)
pheatmap(rd)
```

```{r}
pdb <- read.pdb("1hsg")

plotb3(pdbs$b[1,], typ="l", lwd=2, sse=pdb)
points(pdbs$b[2,], typ="l", col="red")
points(pdbs$b[3,], typ="l", col="blue")
points(pdbs$b[4,], typ="l", col="darkgreen")
points(pdbs$b[5,], typ="l", col="orange")
abline(v=100, col="gray")
```

```{r}
core <- core.find(pdbs)

core.inds <- print(core, vol=0.5)

xyz <- pdbfit(pdbs, core.inds, outpath="corefit_structures")

rf <- rmsf(xyz)

plotb3(rf, sse=pdb)
abline(v=100, col="gray", ylab="RMSF")
```

## 8.2 Predicted Alignment Error for domains

```{r}
#install.packages("jsonlite")

library(jsonlite)

# Listing of all PAE JSON files
pae_files <- list.files(path=results_dir,
                        pattern=".*model.*\\.json",
                        full.names = TRUE)

pae1 <- read_json(pae_files[1],simplifyVector = TRUE)
pae5 <- read_json(pae_files[5],simplifyVector = TRUE)

attributes(pae1)

# Per-residue pLDDT scores 
#  same as B-factor of PDB..
head(pae1$plddt) 

pae1$max_pae

pae5$max_pae

plot.dmat(pae1$pae, 
          xlab="Residue Position (i)",
          ylab="Residue Position (j)")

plot.dmat(pae5$pae, 
          xlab="Residue Position (i)",
          ylab="Residue Position (j)",
          grid.col = "black",
          zlim=c(0,30))

plot.dmat(pae1$pae, 
          xlab="Residue Position (i)",
          ylab="Residue Position (j)",
          grid.col = "black",
          zlim=c(0,30))
```

## 8.3 Residue conservation from alignment file

```{r}
aln_file <- list.files(path=results_dir,
                       pattern=".a3m$",
                        full.names = TRUE)
aln_file

aln <- read.fasta(aln_file[1], to.upper = TRUE)

dim(aln$ali)

sim <- conserv(aln)

plotb3(sim[1:99], sse=trim.pdb(pdb, chain="A"),
       ylab="Conservation Score")

con <- consensus(aln, cutoff = 0.9)
con$seq

m1.pdb <- read.pdb(pdb_files[1])
occ <- vec2resno(c(sim[1:99], sim[1:99]), m1.pdb$atom$resno)
write.pdb(m1.pdb, o=occ, file="m1_conserv.pdb")
```
