---
title: "Lab 19: Pertussis Resurgence"
author: "Nicholas Chiu"
format: pdf
editor: 
  markdown: 
    wrap: 72
---

## Load Package + Data

```{r}
library(datapasta)
library(ggplot2)

cdc <- data.frame(
                                 Year = c(1922L,1923L,1924L,1925L,
                                          1926L,1927L,1928L,1929L,1930L,1931L,
                                          1932L,1933L,1934L,1935L,1936L,
                                          1937L,1938L,1939L,1940L,1941L,1942L,
                                          1943L,1944L,1945L,1946L,1947L,
                                          1948L,1949L,1950L,1951L,1952L,
                                          1953L,1954L,1955L,1956L,1957L,1958L,
                                          1959L,1960L,1961L,1962L,1963L,
                                          1964L,1965L,1966L,1967L,1968L,1969L,
                                          1970L,1971L,1972L,1973L,1974L,
                                          1975L,1976L,1977L,1978L,1979L,1980L,
                                          1981L,1982L,1983L,1984L,1985L,
                                          1986L,1987L,1988L,1989L,1990L,
                                          1991L,1992L,1993L,1994L,1995L,1996L,
                                          1997L,1998L,1999L,2000L,2001L,
                                          2002L,2003L,2004L,2005L,2006L,2007L,
                                          2008L,2009L,2010L,2011L,2012L,
                                          2013L,2014L,2015L,2016L,2017L,2018L,
                                          2019L,2020L,2021L),
         No..Reported.Pertussis.Cases = c(107473,164191,165418,152003,
                                          202210,181411,161799,197371,
                                          166914,172559,215343,179135,265269,
                                          180518,147237,214652,227319,103188,
                                          183866,222202,191383,191890,109873,
                                          133792,109860,156517,74715,69479,
                                          120718,68687,45030,37129,60886,
                                          62786,31732,28295,32148,40005,
                                          14809,11468,17749,17135,13005,6799,
                                          7717,9718,4810,3285,4249,3036,
                                          3287,1759,2402,1738,1010,2177,2063,
                                          1623,1730,1248,1895,2463,2276,
                                          3589,4195,2823,3450,4157,4570,
                                          2719,4083,6586,4617,5137,7796,6564,
                                          7405,7298,7867,7580,9771,11647,
                                          25827,25616,15632,10454,13278,
                                          16858,27550,18719,48277,28639,32971,
                                          20762,17972,18975,15609,18617,
                                          6124,2116)
       )

summary(cdc)

```

## 1. Investigating pertussis cases by year

Q1:

```{r}
caseByYearPlot <- ggplot(cdc) +
  aes(Year, No..Reported.Pertussis.Cases) +
  geom_point() +
  geom_line() +
  labs("Pertussis Cases by Year") + ylab("Number of Cases")

caseByYearPlot
```

## 2. A tale of two vaccines (wP & aP)

Q2:

```{r}
caseByYearPlot + geom_vline(xintercept = c(1946,1996), linetype="dashed", color = c("blue","red"))
```

Q3: After the introduction of the aP vaccine, cases increased again
starting around the 2000s but decreased again 20 years after. One
possible explanation for the observed trend is that there are less
people who are getting the vaccine compared to before. The sensitivity
of the detection of Pertussis could have also increased.

## 3. Exploring CMI-PB data

```{r}
# Allows us to read, write and process JSON data
library(jsonlite)

subject <- read_json("https://www.cmi-pb.org/api/subject", simplifyVector = TRUE) 

head(subject, 3)

table(subject$infancy_vac)

table(subject$biological_sex)

table(subject$biological_sex, subject$race)
```

Q4: There are 58 wP and 60 aP infancy vaccinated subjects in the
dataset.

Q5: 79 female and 39 male subjects.

Q6: 

0 American Indian/Alaska Native Female, 1 American Indian/Alaska
Native Male 

21 Asian Female11 Asian Male 

2 Black Female, 0 Black Male

9 More Than One Race Female, 2 More Than One Race Male 

1 Native Hawaiian or Other Pacific Islander Female, 1 Native Hawaiian or Other Pacific Islander Male

11 Unknown or Not Reported Female, 4 Unknown or Not Reported Male

35 White Female, 20 White Male

Q7: 

From the tables alone, the age groups appear to be significantly different
```{r}
#install.packages("lubridate")
library(lubridate)

subject$age <- today() - ymd(subject$year_of_birth)

library(dplyr)

ap <- subject %>% filter(infancy_vac == "aP")

round( summary( time_length( ap$age, "years" ) ) )

wp <- subject %>% filter(infancy_vac == "wP")

round( summary( time_length( wp$age, "years" ) ) )



```

Q8:
```{r}
int <- ymd(subject$date_of_boost) - ymd(subject$year_of_birth)
age_at_boost <- time_length(int, "year")
head(age_at_boost)
```

Q9:

These two groups are significantly different.
```{r}
ggplot(subject) +
  aes(time_length(age, "year"),
      fill=as.factor(infancy_vac)) +
  geom_histogram(show.legend=FALSE) +
  facet_wrap(vars(infancy_vac), nrow=2) +
  xlab("Age in years")
```


```{r}
library(dplyr)

specimen <- read_json("https://www.cmi-pb.org/api/specimen", simplifyVector = TRUE) 
titer <- read_json("https://www.cmi-pb.org/api/v4/plasma_ab_titer", simplifyVector = TRUE) 
```

Q9.2:

```{r}
meta <- inner_join(specimen, subject)
dim(meta)
head(meta)
```

Q10:

```{r}
abdata <- inner_join(titer, meta)
dim(abdata)
```

Q11:

```{r}
table(abdata$isotype)
```

Q12:

31520, 8085, 2170

The number of rows for the most recent dataset is much smaller than the previous 2. 

```{r}
table(abdata$dataset)
```

## 4. Examine IgG Ab titer levels

```{r}
igg <- abdata %>% filter(isotype == "IgG")
head(igg)
```

Q13:

```{r}
ggplot(igg) +
  aes(MFI_normalised, antigen) +
  geom_boxplot() + 
    xlim(0,75) +
  facet_wrap(vars(visit), nrow=2)
```
Q14:
PT, PRN, FIM2/3, and FHA show differences in the level of IgG antibody titers recognizing them over time. This is likely because they are involved in active infection and virulence as well as cell binding. 

```{r}
ggplot(igg) +
  aes(MFI_normalised, antigen, col=infancy_vac ) +
  geom_boxplot(show.legend = FALSE) + 
  facet_wrap(vars(visit), nrow=2) +
  xlim(0,75) +
  theme_bw()

igg %>% filter(visit != 8) %>%
ggplot() +
  aes(MFI_normalised, antigen, col=infancy_vac ) +
  geom_boxplot(show.legend = FALSE) + 
  xlim(0,75) +
  facet_wrap(vars(infancy_vac, visit), nrow=2)
```

Q15:

```{r}
filter(igg, antigen=="OVA") %>%
  ggplot() +
  aes(MFI_normalised, col=infancy_vac) +
  geom_boxplot(show.legend = F) +
  facet_wrap(vars(visit)) +
  theme_bw() +
  labs(title="OVA antigen levels per visit (aP red, wP teal)")

filter(igg, antigen=="PT") %>%
  ggplot() +
  aes(MFI_normalised, col=infancy_vac) +
  geom_boxplot(show.legend = F) +
  facet_wrap(vars(visit)) +
  theme_bw() +
  labs(title="PT antigen levels per visit (aP red, wP teal)")
```

Q16: 

PT levels increase over time and are much larger than OVA antigen levels. PT levels for both vaccines peaks around the fifth visit and declines while the OVA levels stay relatively constant in comparison.

Q17: When looking at the PT antigen levels, there does not appear to be an extremely significant difference between the two vaccines. From visits 4-6, aP does have a higher median antigen level but not significantly.

```{r}
abdata.21 <- abdata %>% filter(dataset == "2021_dataset")

abdata.21 %>% 
  filter(isotype == "IgG",  antigen == "PT") %>%
  ggplot() +
    aes(x=planned_day_relative_to_boost,
        y=MFI_normalised,
        col=infancy_vac,
        group=subject_id) +
    geom_point() +
    geom_line() +
    geom_vline(xintercept=0, linetype="dashed") +
    geom_vline(xintercept=14, linetype="dashed") +
  labs(title="2021 dataset IgG PT",
       subtitle = "Dashed lines indicate day 0 (pre-boost) and 14 (apparent peak levels)")

abdata.20 <- abdata %>% filter(dataset == "2020_dataset" & planned_day_relative_to_boost<200)
abdata.20 %>% 
  filter(isotype == "IgG",  antigen == "PT") %>%
  ggplot() +
    aes(x=planned_day_relative_to_boost,
        y=MFI_normalised,
        col=infancy_vac,
        group=subject_id) +
    geom_point() +
    geom_line() +
    geom_vline(xintercept=0, linetype="dashed") +
    geom_vline(xintercept=14, linetype="dashed") +
  labs(title="2020 dataset IgG PT",
       subtitle = "Dashed lines indicate day 0 (pre-boost) and 14 (apparent peak levels)")
```
Q18:

The 2020 dataset displays a similar trend to the 2021 dataset.

## 5. Obtaining CMI-PB RNASeq data

```{r}
url <- "https://www.cmi-pb.org/api/v2/rnaseq?versioned_ensembl_gene_id=eq.ENSG00000211896.7"

rna <- read_json(url, simplifyVector = TRUE) 
```

```{r}
#meta <- inner_join(specimen, subject)
ssrna <- inner_join(rna, meta)
```

Q19:

```{r}
ggplot(ssrna) +
  aes(visit, tpm, group=subject_id) +
  geom_point() +
  geom_line(alpha=0.2)
```

Q20:

IGHG1 gene expression increases over time, with a distict peak at visit 4, before decreasing.

Q21:

This pattern does match the trend of ab titer data. At around visit 5 is when antigen levels peak, which is right after the gene is most strongly expressed. Since antigens are long-lived, the levels of antigens remains high while the gene expression decreases.

```{r}
ggplot(ssrna) +
  aes(tpm, col=infancy_vac) +
  geom_boxplot() +
  facet_wrap(vars(visit))

ssrna %>%  
  filter(visit==4) %>% 
  ggplot() +
    aes(tpm, col=infancy_vac) + geom_density() + 
    geom_rug() 
```

